docs/docs/server-middleware.md

```diff
@@ -0,0 +1,649 @@
+Sets response header with the request duration.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.ResponseTiming
+
+val timingService = ResponseTiming(service.orNotFound)
+```
+```scala mdoc
+timingService(okRequest).unsafeRunSync().headers
+```
+
+### RequestId
+
+Use the `RequestId` middleware to automatically generate a `X-Request-ID` header for a request,
+if one wasn't supplied. Adds a `X-Request-ID` header to the response with the id generated
+or supplied as part of the request.
+
+This [heroku guide](https://devcenter.heroku.com/articles/http-request-id) gives a brief explanation
+as to why this header is useful.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.RequestId
+
+val requestIdService = RequestId.httpRoutes(HttpRoutes.of[IO] {
+  case req =>
+    val reqId = req.headers.get(ci"X-Request-ID").fold("null")(_.head.value)
+    // use request id to correlate logs with the request
+    IO(println(s"request received, cid=$reqId")) *> Ok()
+})
+val responseIO = requestIdService.orNotFound(okRequest)
+
+// generated request id can be correlated with logs
+val resp = responseIO.unsafeRunSync()
+```
+Note: `req.attributes.lookup(RequestId.requestIdAttrKey)` can also be used to lookup the request id
+extracted from the header, or the generated request id.
+```scala mdoc
+resp.headers
+resp.attributes.lookup(RequestId.requestIdAttrKey)
+```
+
+### StaticHeaders
+
+Adds static headers to the response.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.StaticHeaders
+
+val staticHeadersService = StaticHeaders(Headers("X-Hello" -> "hi"))(service).orNotFound
+```
+```scala mdoc
+staticHeadersService(okRequest).unsafeRunSync().headers
+```
+
+## Request rewriting
+
+### AutoSlash
+
+Removes a trailing slash from the requested url so that requests with trailing slash map to the route without.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.AutoSlash
+
+val autoSlashService = AutoSlash(service).orNotFound
+val okWithSlash = Request[IO](Method.GET, uri"/ok/")
+```
+```scala mdoc
+// without the middleware the request with trailing slash fails
+service.orNotFound(okRequest).unsafeRunSync().status
+service.orNotFound(okWithSlash).unsafeRunSync().status
+
+// with the middleware both work
+autoSlashService(okRequest).unsafeRunSync().status
+autoSlashService(okWithSlash).unsafeRunSync().status
+```
+
+### DefaultHead
+Provides a naive implementation of a HEAD request for any GET routes. The response has the same headers
+but no body. An attempt is made to interrupt the process of generating the body.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.DefaultHead
+
+val headService = DefaultHead(service).orNotFound
+```
+`/forever` has an infinite body but the `HEAD` request completes without body and includes the headers:
+```scala mdoc
+headService(Request[IO](Method.HEAD, uri"/forever"))
+  .flatMap(r => 
+    r.bodyText.compile.last.map(body => (body, r.status, r.headers))
+  ).unsafeRunSync()
```

Comment:
This is a bit weird.
> A response to a HEAD method should not have a body. If it has one anyway, that body must be ignored

https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD

Reply:
My intent was to show  that even though `GET /forever` produces an infinite body, the HEAD implementation provided by the middleware returns no body (the evaluated renders `None`)

Reply:
Hmm, yeah, I see now 🤔 IMHO we shouldn't put "wrong" code in copy-pastable code blocks.

Maybe it's not worth it, but couldn't we better demo this with e.g. `Client.fromHttpApp(...).status("/forever")` type of thing? That method would only terminate and return if it's not actually draining the forever body, right?

Reply:
Yeah not hanging would be enough demonstration, I think. I haven't used `Client.fromHttpApp`, I guess it would solve the issues around not draining the body. I'll do that for all examples, I can now remove the warning in the beginning and a bunch o `.as[Unit]` .

Reply:
> Yeah not hanging would be enough demonstration, I think. I haven't used `Client.fromHttpApp`, I guess it would solve the issues around not draining the body. I'll do that for all examples, I can now remove the warning in the beginning and a bunch o `.as[Unit]` .

I think Client.status won't pull the body anyway, so it doesn't work to demonstrate the effects of middleware that depends on the request ending. `Client.fromHttpApp` still seems to be better didactically, so I'll continue to write the examples with it.

Reply:
> I think Client.status won't pull the body anyway

It should, if it doesn't then it's a bug :) in a request, the response body always has to be drained. Indeed you can see there is custom logic to handle that:
https://github.com/http4s/http4s/blob/5f962a6d64658e7f8f272b2fe1aa58c2f0fd20fc/client/shared/src/main/scala/org/http4s/client/Client.scala#L292-L299

Reply:
Hmm, reading that implementation more carefully, it actually seems like it's _not_ draining the body. So that does seem like a bug 🤔

Reply:
I have one or two examples where i use `.expect[Unit]` (and if were to use `.status(...).void` they wouldn't work because of the bug above)  but they still make sense because I don't need anything from the response.